File: src/leetcode/python/ProductOfArrayExceptSelf_238.py
class ProductOfArrayExceptSelf_238(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        l = len(nums)
        left_product = [1] * l
        for i in range(1, l):
            left_product[i] = left_product[i-1] * nums[i-1]
        right_product = 1
        for i in range(l-1, -1, -1):
            left_product[i] = left_product[i] * right_product
            right_product *= nums[i]
        return left_product

File: src/leetcode/java/ProductOfArrayExceptSelf_238.java
package leetcode.java;

public class ProductOfArrayExceptSelf_238 {
  public int[] productExceptSelf(int[] nums) {
    int[] leftProduct = new int[nums.length];
    leftProduct[0] = 1;
    for (int i = 1; i < nums.length; i++) {
      leftProduct[i] = leftProduct[i - 1] * nums[i - 1];
    }
    int rightProduct = 1;
    for (int i = nums.length - 1; i >= 0; i--) {
      leftProduct[i] *= rightProduct;
      rightProduct *= nums[i];
    }
    return leftProduct;
  }
}
